src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/authorization/ServiceProviderAccessScopeMapper.kt

```diff
@@ -40,36 +40,51 @@ class ServiceProviderAccessScopeMapper(
       .filter { it.startsWith(serviceProviderGroupPrefix) }
       .map { it.removePrefix(serviceProviderGroupPrefix) }
 
-    if (serviceProviderGroups.isEmpty()) {
-      configErrors.add("no service provider groups associated with user")
-    } else if (serviceProviderGroups.size > 1) {
-      configErrors.add("more than one service provider groups associated with user")
-    }
-
-    val serviceProvider = serviceProviderRepository.findByIdOrNull(serviceProviderGroups[0])
-
-    if (serviceProvider == null) {
-      configErrors.add("service provider id does not exist in the interventions database")
-    }
+    val serviceProvider = getServiceProvider(serviceProviderGroups, configErrors)
 
     val contractGroups = groups
       .filter { it.startsWith(contractGroupPrefix) }
       .map { it.removePrefix(contractGroupPrefix) }
 
-    if (contractGroups.isEmpty()) {
-      configErrors.add("no contract groups associated with user")
-    }
-
-    val contracts = dynamicFrameworkContractRepository.findAllByContractReferenceIn(contractGroups)
-
-    if (contracts.size != contractGroups.size) {
-      configErrors.add("at least one contract group does not exist in the interventions database")
-    }
+    val contracts = getContracts(contractGroups, configErrors)
 
     if (configErrors.isNotEmpty()) {
       throw AccessError(errorMessage, configErrors)
     }
 
     return ServiceProviderAccessScope(serviceProvider = serviceProvider, contracts = contracts)
   }
+
+  private fun getServiceProvider(serviceProviderGroups: List<String>, configErrors: MutableList<String>): ServiceProvider? {
+    return when {
+      serviceProviderGroups.isEmpty() -> {
+        configErrors.add("no service provider groups associated with user")
+        null
+      }
+      serviceProviderGroups.size > 1 -> {
+        configErrors.add("more than one service provider group associated with user")
```

Comment:
More of a question for me... I suppose it's not known how hmpps auth is set up in prod but do we know for certain that users will only ever be set up for a single service provider auth group? 
Is it possible that hmpps auth is not entirely clean - maybe on renaming of an SP they don't delete the old? I know it's not our responsibility - but is there some way to know about the cleanliness of hmpps auth before prod live to reduce issues arising?

Reply:
great question - this is a constraint that we are in charge of. since we have no idea about which is the correct provider in the case where there is more than one, we have to enforce this rule.

i'm sure we could run queries against auth to validate that there is only one INT_SP_ group per user, but really this is more of a business rule around 'onboarding' SP users into auth. we can raise this with sarah smith to discuss further.